consensus/core/src/leader_scoring.rs

```diff
@@ -138,173 +69,203 @@ impl ReputationScores {
+                        continue;
+                    }
+
+                    // If a blocks strong linked ancestor is in leaders, then
+                    // it's a vote for leader.
+                    if self.leaders.contains(ancestor) {
+                        // There should never be duplicate references to blocks
+                        // with strong linked ancestors to leader.
+                        tracing::trace!(
+                            "Found a vote {} for leader {ancestor} from authority {}",
+                            block.reference(),
+                            block.author()
+                        );
+                        assert!(self
+                            .votes
+                            .insert(block.reference(), StakeAggregator::new())
+                            .is_none(), "Vote {block} already exists. Duplicate vote found for leader {ancestor}");
+                    }
+
+                    if let Some(stake) = self.votes.get_mut(ancestor) {
+                        // Vote is strongly linked to a future block, so we
+                        // consider this a certified vote.
+                        tracing::trace!(
+                            "Found a certified vote {ancestor} from authority {}",
+                            ancestor.author
+                        );
+                        stake.add(block.author(), &self.context.committee);
+                    }
+                }
             }
-            if let Some(ancestor) = self.get_block(ancestor) {
-                if let Some(support) = self.find_supported_leader_block(leader_slot, &ancestor) {
-                    return Some(support);
+        }
+    }
+
+    // Iterate through votes and calculate scores for each authority based on
+    // scoring strategy that is used. (Vote or CertifiedVote)
+    pub(crate) fn calculate_scores(&self) -> ReputationScores {
+        let _scope = mysten_metrics::monitored_scope("ScoringSubdag::calculate_scores");
+
+        let scores_per_authority = if self
+            .context
+            .protocol_config
+            .consensus_certified_vote_scoring_strategy()
+        {
+            if let Ok(scoring_strategy) = std::env::var("CONSENSUS_SCORING_STRATEGY") {
+                if scoring_strategy == "certified_vote_v3" {
+                    self.score_certified_votes_v3()
+                } else {
+                    self.score_certified_votes_v2()
                 }
             } else {
-                // TODO: Add unit test for this case once dagbuilder is ready.
-                tracing::trace!(
-                    "Potential vote's ancestor block not found in unscored committed subdags: {:?}",
-                    ancestor
-                );
-                return None;
+                self.score_certified_votes_v2()
             }
+        } else {
+            self.score_votes()
+        };
+
+        // TODO(arun): Normalize scores
```

Comment:
Wanted to add the normalizing of scores from 0-1 but realized that would mean we would need to change ReputationScores to hold a vec<f64> instead of vec<u64> which then means we have to change what is stored in RocksDB.